docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,395 @@
+from the outside. Conversely, if we allow a token program to take the lamports it wants,
+a malicious token program could take too much from users.
+
+To get around this, the Token Program interface has a new `CreateHolding`
+instruction, which only allocates space and assigns the account to the program.
+Once the space is allocated, the Associated Token Program transfers the
+required lamports, and returns an error if that number is too large. The process
+becomes:
+
+* call `CreateHolding` on the token program (allocate and assign)
+* calculate rent requirement based on the data size of holding account
+* `SystemInstruction::Transfer` the required lamports to make the holding account rent-exempt
+* call `InitializeHolding` on the token program
+
+### Web3 / Wallets
+
+To properly support i-tokens, wallets and applications will need to make the
+following changes:
+
+* validate program: using the token-list, check that the given program id is legit
+* get holdings: query for holdings owned by the public key, for all official
+token programs listed in the registry (not just `Tokenkeg...`)
+* get balances: avoid deserializing holding data, and instead use
+`getBalance` from RPC
+* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
+program
+* create holding: avoid directly creating an account and initializing,
+and use the associated token account program instead
+
+### On-chain Programs
+
+To properly support i-tokens on-chain, programs will need to make these changes:
+
+* multiple token programs: all programs that support SPL Tokens must always
+include the program id of the token(s) holdings they're passing in.
+For example, token-swap needs to accept separate token programs for token A and B
+* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
+the SPL token code
+* get holding data: instead of deserializing the account data into an SPL holding, use
+the new wrapper to directly deserialize or call the `NormalizeHolding`
+instruction with the appropriate token program
+* get mint data: same as above, but using the `NormalizeMint`
+* transfer: only use `TransferChecked`, which requires passing mints
+
+### Ledger
+
+Ledger integrates with token-list in order to provide verified token program ids
+and token tickers to the Ledger SDK and eliminate the need to store static
+information in the Ledger Solana App.
+
+A wallet or CLI with Ledger integration fetches the token-list and signing
+information. When the user wishes to transfer tokens, the wallet provides
+Ledger with the signed token and program metadata.
+
+The Solana App pushes a signature verification on the metadata using Ledger's
+native transaction model. If the signature verification passes, then the program
+id and token mint information are valid, and the actual transfer proceeds. If
+it fails due to spoofed program or mint data, the whole transaction aborts.
+
+TODO any associated token program considerations?
```

Comment:
This should be compatible with support for the current ATA implementation, so long as the instruction format doesn't change